File: src/main/scala/laoflch/debezium/connector/informix/InformixEventMetadataProvider.scala
package laoflch.debezium.connector.informix.integrtest

import java.time.Instant
import java.util
import java.util.Map

import org.apache.kafka.connect.data.Struct
import io.debezium.data.Envelope
import io.debezium.pipeline.source.spi.EventMetadataProvider
import io.debezium.pipeline.spi.OffsetContext
import io.debezium.schema.DataCollectionId
import io.debezium.util.Collect


class InformixEventMetadataProvider extends EventMetadataProvider {

  override def getEventTimestamp(source: DataCollectionId, offset: OffsetContext, key: Any, value: Struct): Instant = {
    if (value == null) return null
    val sourceInfo = value.getStruct(Envelope.FieldName.SOURCE)
    if (source == null) return null
    val timestamp = sourceInfo.getInt64(SourceInfo.TIMESTAMP_KEY)
    if (timestamp == null) null
    else Instant.ofEpochMilli(timestamp)
  }

  override def getEventSourcePosition(source: DataCollectionId, offset: OffsetContext, key: Any, value: Struct): util.Map[String, String] = {
    if (value == null) return null
    val sourceInfo = value.getStruct(Envelope.FieldName.SOURCE)
    if (source == null) return null
    Collect.hashMapOf(/*SourceInfo.COMMIT_LSN_KEY, sourceInfo.getString(SourceInfo.COMMIT_LSN_KEY),*/ SourceInfo.CHANGE_LSN_KEY, sourceInfo.getString(SourceInfo.CHANGE_LSN_KEY))
  }

  override def getTransactionId(source: DataCollectionId, offset: OffsetContext, key: Any, value: Struct): String = {
    if (value == null) return null
    val sourceInfo = value.getStruct(Envelope.FieldName.SOURCE)
    if (source == null) return null
    sourceInfo.getString(SourceInfo.TX_ID)
  }
}

File: src/main/java/io/debezium/connector/informix/InformixEventMetadataProvider.java
/*
 * Copyright Debezium Authors.
 *
 * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 */
package io.debezium.connector.informix;

import java.time.Instant;
import java.util.Map;

import org.apache.kafka.connect.data.Struct;

import io.debezium.data.Envelope;
import io.debezium.pipeline.source.spi.EventMetadataProvider;
import io.debezium.pipeline.spi.OffsetContext;
import io.debezium.spi.schema.DataCollectionId;
import io.debezium.util.Collect;

public class InformixEventMetadataProvider implements EventMetadataProvider {

    @Override
    public Instant getEventTimestamp(DataCollectionId source, OffsetContext offset, Object key, Struct value) {
        if (value == null || source == null) {
            return null;
        }
        final Long timestamp = value.getStruct(Envelope.FieldName.SOURCE).getInt64(SourceInfo.TIMESTAMP_KEY);
        return timestamp == null ? null : Instant.ofEpochMilli(timestamp);
    }

    @Override
    public Map<String, String> getEventSourcePosition(DataCollectionId source, OffsetContext offset, Object key, Struct value) {
        if (value == null || source == null) {
            return null;
        }
        final Struct sourceInfo = value.getStruct(Envelope.FieldName.SOURCE);
        return Collect.hashMapOf(
                SourceInfo.COMMIT_LSN_KEY, sourceInfo.getString(SourceInfo.COMMIT_LSN_KEY),
                SourceInfo.CHANGE_LSN_KEY, sourceInfo.getString(SourceInfo.CHANGE_LSN_KEY),
                SourceInfo.BEGIN_LSN_KEY, sourceInfo.getString(SourceInfo.BEGIN_LSN_KEY));
    }

    @Override
    public String getTransactionId(DataCollectionId source, OffsetContext offset, Object key, Struct value) {
        return value == null || source == null ? null
                : value
                        .getStruct(Envelope.FieldName.SOURCE)
                        .getString(SourceInfo.TX_ID_KEY);
    }
}
